### ISSUE TITLE
💬 RFC: Dynamic backend plugins

### ISSUE THREAD

**Issue by davidfestal** [backstage/backstage#18390]
### 🔖 Need

As a Backstage adopter, I find it hard to start using or even just quickly try new plugins. Installing plugins is a manual process that requires code changes.

It should be possible to install a backend plugin inside a backstage backend application **without having to modify and recompile the backend application**.

For this purpose, a backstage application should be able to:
- discover plugin packages that should be installed,
- load them as node modules and initialize them in a similar way as what is done for statically-compiled plugins.

To summarize, a backend application should be able to **dynamically load backend plugins**, on condition that they would be appropriately packaged.

In order to make it possible, it should be simple to enhance and repackage an existing plugin in order to make it compatible with dynamic loading. Supporting such dynamic loading should not require existing plugins to be rewritten, and be strictly optional.

It would be acceptable that a backend application would need to be restarted in order to add / remove / update a dynamic plugin. Plugin installation/update/removal while the backend application is already running is out of the scope of this proposal.

#### Scope

The scope of the proposal is limited to loading and initializing plugins available as local package folders.
*The question of how these packages would be bundled, exposed, distributed or downloaded is on-purpose out-of-scope of this proposal.*

#### Guidelines

The proposal presented here is based on the following guidelines:
1. A dynamic plugin should be loaded through a standard dynamic `import(<path>)` call, which would return a module containing all the exported definitions of the related backstage plugin.
2. To make this loading both possible and robust, the plugin should be repackaged to be **standalone**, and **self-contained**, providing its specific dependencies
3. As a special case, it should never provide any core `@backstage` dependency. On the contrary, it should always **share `@backstage` dependencies** with the backend application (to avoid having dupes of `@backstage` modules)
4. A dynamic plugin should be loaded from either the old backend system, or the new backend system, according to the context, and provide contribution to the main types of contributions (router, catalog, scaffolder, search, events, permissions) for both systems.
5. Making a plugin compatible should require as **minimal and non-invasive source code changes** as possible.
6. It should be possible to repackage an existing backend plugin as a dynamic plugin **without requiring any change** in the existing plugin source code.
7. The dynamic backend plugin discovery and loading should require as less configuration as possible.
8. The dynamic backend plugin feature should be optional and should not change the way static plugins work today.

#### Proposed API

The backstage backend should be able to manage the concept of dynamic plugin which boils down to 3 main API entities:
- a [**scanned plugin package**](https://github.com/davidfestal/backstage/blob/97de141cf75e49f2bb9863bc40f119e6c91398f4/packages/backend-plugin-manager/src/scanner/types.ts#L19-L22), discovered at a given url, and providing a `package.json` manifest:

```typescript
export interface ScannedPluginPackage {
  location: URL;
  manifest: ScannedPluginManifest;
}
```

- a [**dynamic plugin**](https://github.com/davidfestal/backstage/blob/97de141cf75e49f2bb9863bc40f119e6c91398f4/packages/backend-plugin-manager/src/manager/types.ts#L62-L76) (either frontend or backend) featuring its name, version, role, and platform (`web` or `node`), and, for backend dynamic plugins an **installer**:

```typescript
export interface DynamicPlugin {
  name: string;
  version: string;
  role: PackageRole;
  platform: PackagePlatform;
}

export interface FrontendDynamicPlugin extends DynamicPlugin {
  platform: 'web';
}

export interface BackendDynamicPlugin extends DynamicPlugin {
  platform: 'node';
  installer: BackendDynamicPluginInstaller;
}
```

> ##### Note:
> Though this RFC is focused towards implementing the support for dynamic **backend** plugins, front-end plugins have been considered as a discoverable entity api-wise because frontend and backend plugins generally go hand-in-hand to comprise a single feature. We can then envision that, in future steps of the dynamic plugins work, whatever be the way dynamic frontend plugins are implemented and wired into the front-end application, it would make sense for the plugin manager running on the backend to be able to discover them.
 

- a [**dynamic backend plugin installer**](https://github.com/davidfestal/backstage/blob/97de141cf75e49f2bb9863bc40f119e6c91398f4/packages/backend-plugin-manager/src/manager/types.ts#L78-L110), which contains the logic required to initialize the dynamic backend plugin (specific to either the `legacy` or the `new` backend system):

```typescript
export type BackendDynamicPluginInstaller =
  | LegacyBackendPluginInstaller
  | NewBackendPluginInstaller;

export interface NewBackendPluginInstaller {
  kind: 'new';
  install(): BackendFeature | BackendFeature[];
}

export interface LegacyBackendPluginInstaller {
  kind: 'legacy';
  router?: {
    pluginID: string;
    createPlugin(env: PluginEnvironment): Promise<Router>;
  };
  catalog?(builder: CatalogBuilder, env: PluginEnvironment): void;
  scaffolder?(env: PluginEnvironment): TemplateAction<any>[];
  search?(
    indexBuilder: IndexBuilder,
    schedule: TaskRunner,
    env: PluginEnvironment,
  ): void;
  events?(
    eventsBackend: EventsBackend,
    env: PluginEnvironment,
  ): HttpPostIngressOptions[];
  permissions?: {
    policy?: PermissionPolicy;
  };
}
```

In order for a backend plugin to be compatible with dynamic loading, it should **export such an installer object** (either a legacy or a new one, or both according to the context) as a constant named **with a fixed name** through which it will be loaded (for example `dynamicPluginInstaller`).

#### New service

A new service, `backend plugin manager` would:
- scan a single given root folder (configured in the `app.config`) to search for sub-folders with `package.json` files, and build a list of **scanned plugin packages**,
- import every appropriate module (legacy or new backend plugin),
- retrieve the appropriate installer for every imported module,
- expose the list of **dynamic plugins**, along with their installer so that:
  - the main backend application can initialize them (new backend system) or start the corresponding router (legacy backend system),
  - contribution-receiving plugins (catalog, search, scaffolder, ...) can install the dynamic plugin contributions (legacy backend system).

#### New CLI command

A new CLI command (`export-dynamic` for example) would allow exporting a backend plugin as a package prepared for dynamic loading. A package exported for dynamic loading differs from the original package content by the following:
- the javascript code is still compiled the same way as before, but might embed some modules from the direct dependencies in the output code, to have the dynamic plugin more easily self-contained (typical use-case for this is the case of sibling `xxx-common` source packages).
- the `package.json` is modified to:
  - suffix the package name with `-dynamic` to avoid confusion at publishing time,
  - hoist dependencies of embedded modules as direct dependencies,
  - fix version of `@backstage` dependencies and move them to peer dependencies (because we expect them being already loaded by the backstage backend app)
  - set `bundleDependencies` to `true`, in order to  allow the modified package to be fully self-contained (with the `node_modules` filled) when published through NPM publish.

The dynamic plugin package would be exported in a new sub-folder of the original package (`dist-dynamic` for example), just besides the `dist` folder, and would be integrated in the original package distributed files.

The CLI command would run `yarn install` in the newly-created `dist-dynamic` folder, so that a fixed `yarn.lock` would be part of the exported package, which is also important to have it really self-contained and immutable.

##### Legacy backend system

The wiring of dynamic plugin loading inside the legacy backend application mainly consists in iterating through the plugins exposed by the new backend plugin manager service, and applying their contributions at the right place:
- the contributions to router would be applied in the main `index.ts` file of the backend application
- the contributions to other basic features like search, scaffolder, etc ... would be applied in the feature-related source file (`plugins/search.ts, plugin/scaffolder.ts`, etc ...).

So enabling the dynamic plugin support in a legacy backend application boils down to changing the application template or providing an alternate one. It would have zero impact on existing applications that do not explicitly update the backend application code to the alternate template.

##### New backend system

The wiring of the dynamic plugin loading inside the new backend system would mainly require a change in the [`BackendInitializer`](https://github.com/backstage/backstage/blob/0f02bdcbd72d09411bbf0ca1dfbf536a35cd2e24/packages/backend-app-api/src/wiring/BackendInitializer.ts#L139): if the backend plugin manager service would be in the available services, it should be used, after all the services are initialized, to retrieve the additional features of the dynamic plugins or modules, and add them in the list of features before proceeding to the feature initialization step.

In order to allow contributing to extensions exposed by dynamic backend plugins, it would also require changing, in the same file, [the `#extensionPoints` map](https://github.com/backstage/backstage/blob/0f02bdcbd72d09411bbf0ca1dfbf536a35cd2e24/packages/backend-app-api/src/wiring/BackendInitializer.ts#L43), so that its keys would be extension point IDs (string), instead of being extension point refs (objects).

These are quite **simple and non-invasive changes**, especially if the backend plugin manager service is not provided by default.

Then, enabling the dynamic plugin support in a new backend application would boil down to adding the backend plugin manager service factory in the list of services passed to the `createBackend()` call in the main backend application file, like [this](https://github.com/backstage/backstage/commit/da8fb3f925181111407ccd6351485a6d4cadde8b#diff-2f88293777ab85370fa9ae0162a633a27145a55389a43c0b1b8c9b0a7b5ccad9L35-R38):

```diff
- const backend = createBackend();
+ const backend = createBackend({
+   services: [dynamicPluginsServiceFactory()],
+ });
```

It would have zero impact on existing applications that do not explicitly update the backend application code.

#### Iterative implementation

In fact enabling dynamic plugins would require very few changes to the existing backstage code and logic:
- `BackendInitializer` changes mentioned above,
- possibly some optional way to customize the content of a `package.json` in the [CLI production packaging library](https://github.com/backstage/backstage/blob/0f02bdcbd72d09411bbf0ca1dfbf536a35cd2e24/packages/cli/src/lib/packager/productionPack.ts#L34).

All the rest could be new code, and alternate backend application templates.

That would make it possible to introduce an experimental support for dynamic plugins very quickly in the bactstage code, as an optional and experimental feature, without impacting at all the existing behavior of any backstage application.

Proceeding in such a way would provide room for feedback and enhancements, before moving forward and enabling dynamic plugins by default in Backstage applications.

### 〽️ Alternatives

- Decoupling dynamic plugins into a microservices architecture, where each dynamic plugin is a separate microservice. The cost of change of the implementation of such an alternate design would probably be huge, as it would be a full blown rearchitecture, but still it is an alternative.

### ❌ Risks

- Higher memory consumption due to duplicate dependencies of dynamic plugins, since every dynamic plugin would be self-contained and would bring its own private dependencies.
- Since it relies on importing Javascript code at runtime (at NodeJS backend application start, more precisely), dynamic plugins would make it easier to :
    - include arbitrary code that might cause failure or unexpected behavior of the backend app.
    - add a plugin with a conflicting dependency version, which could introduce bugs at runtime: this makes various plugin combinations hard to test and support.
    - introduce database inconsistencies if a plugin would try to migrate database model, and then would be removed or downgraded for example.

  In fact the 3 risks mentioned just above are not really specific to dynamic plugins, but such risks are now distributed across both the main backstage application and the various dynamic plugins that would be added. So applying security checks at backstage backend application compile time would not be sufficient to mitigate them. Mitigating them would then probably involve:
    - some sort of dynamic plugins signing that would guarantee that a dynamic plugin is safe,
    - version compatibility rules and safeguards when loading dynamic plugins,
    - package installation hooks allowing for DB migration and other pre-post install/uninstall/upgrade chores. This is out of the scope of this RFC. 

### 👀 Have you spent some time to check if this RFC has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

**Comment by davidfestal** [comment]
> - Are there any plans for this (or follow up work) to provide a "Plugin registry" that can be queried for installed plugins?

Having a way to:
- query the list of currently loaded dynamic plugins and the list of dynamic plugins available for installation,
- consequently detect the possible need for a restart (to load /  update newly-available dynamic plugins), and notify about it

is effectively in the plans of the implementation work that follows this RFC.

> - Are there any plans for plugins to expose "capabilities"?
> - Are there any plans for consumers to lookup plugins and invoke a capability?

I think this would be at a higher level that what this RFC is about. But surely something that could be built on top of the basic dynamic plugin loading mechanism proposed here.

**Comment by Rugvip** [comment]
Thank you for picking up and driving discussion in this area! 🎉 🚀 

Wanna start off with a couple of high-level thoughts on the area before diving into more detailed comments.

I'm gonna claim that dynamic plugins can be split into four major pieces:

- Packaging - given any type of Backstage package, package it into an artifact that can be loaded at runtime
- Distribution - safely distribute and deliver artifacts from the build environment to runtimes
- Loading - enumerate and load available artifacts as runtime JavaScript modules
- Integration - select and configure features provided by the loaded JavaScript modules

I think each of these really can be their own RFC, although perhaps it makes sense to merge packing and loading. I feel distribution, which you mention a little bit, is something we should be able to quite easily leave for later. In particular leaving the exact loader implementation as an optional service that can be replaced makes it much easier to experiment in this area.

I feel that there's quite a lot going on in this RFC, and we'd benefit from breaking down the problem space a bit into pieces that we tackle separately. In particular I feel it best to keep the design effort around backend plugin integration separate, closely tied to #18372, and built on top of #11611. I do feel quite strongly that we should not consider bringing the legacy backend system along in this space, with that effort instead spent towards migrating existing plugins to the new backend system.

Now over to more direct comments on parts of the RFC 😁 

> ### 🔖 Need

I more or less agree with everything in this section 👍 

> ### 🎉 Proposal
> #### Scope
> The scope of the proposal is limited to loading and initializing plugins available as local package folders. _The question of how these packages would be bundled, exposed, distributed or downloaded is on-purpose out-of-scope of this proposal._

You claim this, but then later on you discuss how packaging and distribution is supposed to work through NPM? 😅 I feel it's fine to loosely discuss different distribution models without diving deeply into any of them. One assumption that we could just stop at for now is that a packaged backend plugin artifact is a folder structure with a package.json? On top of that we can then for example do things like package and distribute as tarball, signing the contents, re-using more or less of NPM infra, etc.


> 1. A dynamic plugin should be loaded through a standard dynamic `import(<path>)` call, which would return a module containing all the exported definitions of the related backstage plugin.

Something we'll want to figure out here is the best way to access exports of said module. That goes for frontend as well, so a problem we should just solve once.

> 2. To make this loading both possible and robust, the plugin should be repackaged to be **standalone**, and **self-contained**, providing its specific dependencies

Just wanna highlight that native dependencies are gonna be an annoyance that we need to figure out how to handle. My initial thinking here is that we make them external dependencies that plugins expect to be installed directly in the backend. To support multiple versions of the same package we could allow aliasing with overrides.

> 3. As a special case, it should never provide any core `@backstage` dependency. On the contrary, it should always **share `@backstage` dependencies** with the backend application (to avoid having dupes of `@backstage` modules)

Why? 😅 All relevant libraries should be built so support duplication, and I expect the cost of duplication here to be minimal. I think this is something we can leave as a potential future optimization.

> 4. A dynamic plugin should be loaded from either the old backend system, or the new backend system, according to the context, and provide contribution to the main types of contributions (router, catalog, scaffolder, search, events, permissions) for both systems.

Would you be alright with limiting this and future discussion to just the new backend system? We'd really want to avoid inventing an additional intermediate mini backend system.

> 5. Making a plugin compatible should require as **minimal and non-invasive source code changes** as possible.
> 6. It should be possible to repackage an existing backend plugin as a dynamic plugin **without requiring any change** in the existing plugin source code.

I assume you mean compared to using it as a regular backend plugin right? As in point 5 has already been done.

> #### Proposed API

To be honest, given a design where we have an optional experimental root service that is called to load additional plugins, I don't think this it too big of a deal. As in, we don't really need to commit to a stable API at this point and can kinda just :shipit:, i.e. I think this is a discussion that we can have in a PR.

> ```ts
> export interface NewBackendPluginInstaller {
>   kind: 'new';
>   install(): BackendFeature | BackendFeature[];
> }
> ```

I feel there's a step missing here where the integrations select which exact features they want to actually install. I don't think it's a safe assumption that all available features from a backend plugin should always be installed. I do also think there's an opportunity to think a bit more about how we configure individual plugins and features too while we're at it.

For example, these features might have IDs and that need to be referenced to enable and maybe configure them. Although there might also be particular features that are always installed by default, similar to what we're suggesting in #18372. We've actually had a couple of discussions around either renaming backend features to extensions, or frontend extensions to features, there's a section about that in #18372.

> #### New CLI command

This is an interesting and important bit, I'm torn on whether this is something we should leave for later or dive deeper into. I do somewhat feel we need to do either of those two. For example, when in this NPM distribution example does package installation actually happen? do we intend for these packages to be downloaded directly from NPM or NPM-compatible repository at runtime? If so, do we really need separate packages or could we just download the tarball manually and ignore the dependencies? Quite a couple of questions popping up for me here 😅 

> ##### New backend system

Don't worry about the changes needed here, that can all be done 😁. The `#extensionPoints` map change is something we can do straight away, it's already how the service registry works, just an oversight. It also wouldn't be a bit deal to introduce an experimental service at this point for loading in additional features. The tricky bit is the declarative integration and how that configuration works.

> * Decoupling dynamic plugins into a microservices architecture, where each dynamic plugin is a separate microservice. The cost of change of the implementation of such an alternate design would probably be huge, as it would be a full blown rearchitecture, but still it is an alternative.

This is already the case. Backstage plugins are microservices, you can deploy them separately, just gotta keep the modules along. With the introduction of dynamic backend plugins this is something that we'll want to lean into, ideally it's really simple to set up a cluster of Backstage services all running separate plugins.

I get what you mean here too though, and forcing Backstage plugins to always be deployed as microservices is ofc something we want to avoid.

> ### ❌ Risks

On security risks: +1, in particular we need to be careful with what methods there are for loading remove code. For example I would be happy if it was not possible to execute malicious code even if an attacker is able to inject additional config.

**Comment by Rugvip** [comment]
We spent some time this week working towards the design of backend declarative integration (backend equivalent of #18372). What we found is that we do not need any particular additional configuration other than a declarative method to select which packages you want to load, with the assumption that all features exported by those packages will be picked up and installed automatically.

We're continuing the work by adding support for the design that you proposed through the `dynamicPluginsServiceFactory()`. Most likely it'll be an async loader of `BackendFeature`s. Our initial implementation will be targeted towards integrating packages that are already installed, but it's very much going to allow for you to experiment with dynamically loading packages at runtime as well! 🎉 

Assuming all goes as planned we'll effectively consider declarative integration for backend a solved problem. There are a few more bits that we need ship as well, most importantly we're planning to remove the possibility to pass options to backend features through code, e.g. `catalogPlugin({ processingInterval: 30 })`, that'd have to be config instead - we didn't find any meaningful way to work that into the system. We also plan to make service factories also implement `BackendFeature` so that they can be installed declaratively and dynamically.

Hoping all of that sounds good! Let us know if you have any concerns with that direction.

**Comment by davidfestal** [comment]
> We spent some time this week working towards the design of backend declarative integration (backend equivalent of #18372). What we found is that we do not need any particular additional configuration other than a declarative method to select which packages you want to load, with the assumption that all features exported by those packages will be picked up and installed automatically.

I agree with this.

> 
> We're continuing the work by adding support for the design that you proposed through the `dynamicPluginsServiceFactory()`. Most likely it'll be an async loader of `BackendFeature`s. Our initial implementation will be targeted towards integrating packages that are already installed, but it's very much going to allow for you to experiment with dynamically loading packages at runtime as well! tada

The following Draft PR (that will hopefully be available for review very soon), implements such a loading of `BackendFeature`s in the context of  dynamic plugins: #18862

> 
> Assuming all goes as planned we'll effectively consider declarative integration for backend a solved problem. There are a few more bits that we need ship as well, most importantly we're planning to remove the possibility to pass options to backend features through code, e.g. `catalogPlugin({ processingInterval: 30 })`, that'd have to be config instead - we didn't find any meaningful way to work that into the system.

That makes sense, though maybe it would be interesting to keep both options: allow providing these options  through both configuration and code. Would it make sense ? I'm also thinking at things like the  custom [KubernetesClustersSupplier](https://backstage.io/docs/reference/plugin-kubernetes-backend.kubernetesclusterssupplier): How would we pass this through configuration ? Or would this type of options always be managed through extensions ?

> We also plan to make service factories also implement `BackendFeature` so that they can be installed declaratively and dynamically.

:tada: 

> 
> Hoping all of that sounds good! Let us know if you have any concerns with that direction.

All this seems very interesting. Looking forward to see this go hand in hand with the implementation work already done on dynamic backend plugins (cf PR above).

**Comment by Rugvip** [comment]
Quick reply to keep the discussion going :grin:

> > > 1. A dynamic plugin should be loaded through a standard dynamic `import(<path>)` call, which would return a module containing all the exported definitions of the related backstage plugin.
> > 
> > Something we'll want to figure out here is the best way to access exports of said module. That goes for frontend as well, so a problem we should just solve once.
> 
> Why do you think we should need yet another way to access exports? The entrypoint serves that purpose. I wouldn’t connect this to the frontend system at all, loading additional assets over network is not even remotely related to loading code into the NodeJS runtime.

What I had in mind is how to select specific exports in configuration, or standardize on some specific export. In our experimentation what we're leaning towards at this point is to consider all exports of a module, filter out the ones that are backend features, and then install those. It's straightforward and doesn't require any changes to existing packages that already support the new backend system. On top of that we can declare ways to exclude or only include select exports from a package if that's needed as an escape hatch.

> > > 2. To make this loading both possible and robust, the plugin should be repackaged to be **standalone**, and **self-contained**, providing its specific dependencies
> > 
> > 
> > Just wanna highlight that native dependencies are gonna be an annoyance that we need to figure out how to handle. My initial thinking here is that we make them external dependencies that plugins expect to be installed directly in the backend. To support multiple versions of the same package we could allow aliasing with overrides.
> 
> That doesn't seem realistic to me in terms of user experience, if I understood correctly what you mean here. Allowing such overrides in very specific cases, when the isolation of dynamic plugin dependencies isn't sufficient, could be useful, sure. But requiring the end-user to install all dynamic-plugin-specific dependencies himself, and provide such overrides to resolve himself all the possible version conflicts, doesn't seem to fit well with the user experience expected from the dynamic plugins feature.

Native modules as in ones that use [Node.js' C++ APIs](https://nodejs.org/api/addons.html). There's quite few of these but they may require explicit installation as opposed to just being shipped in a JS code bundle. One existing example is the [`isolated-vm`](https://www.npmjs.com/package/isolated-vm), which in addition to its own installation also for example requires the following packages to be installed on Ubuntu: `python g++ build-essential`.

It's just a technical issue, and if we can figure out a way to handle them in a way that doesn't require additional steps that's ofc preferable.

> > > 3. As a special case, it should never provide any core `@backstage` dependency. On the contrary, it should always **share `@backstage` dependencies** with the backend application (to avoid having dupes of `@backstage` modules)
> > 
> > 
> > Why? sweat_smile All relevant libraries should be built so support duplication, and I expect the cost of duplication here to be minimal. I think this is something we can leave as a potential future optimization.
> 
> There are always dependencies which are required to be singleton. E.g. a logger can’t open 2 file descriptors for the same log file. More generally, if we think of how runtime plugins would work in an extensible framework architecture, it seems to me that it makes sense to distinguish **plugin-specific private dependencies**, from **shared dependencies provided by the framework**. In the context of this RFC, I think it would make sense to consider **backstage dependencies as provided by the framework** by default. In NodeJS terms, this would mean: make them peer dependencies, and ensure that their packages are loaded from the same location as when they are loaded by the main backstage application.
> 

This is already a problem we have to solve for for backends that are integrated through code. Un-solving it for dynamic plugins seems unnecessary. When there is a strict requirement for a singleton it can be installed as an explicit service factory. Forcing sharing of certain dependencies adds a lot of complexity, because it makes us much more limited in our ability to remove exports. For example, with multiple versions of `@backstage/backend-plugin-api` allowed, we can remove or change exports such as `createBackendPlugin` in newer versions, but still keep support for older exports in `@backstage/backend-app-api`.

> > > 4. A dynamic plugin should be loaded from either the old backend system, or the new backend system, according to the context, and provide contribution to the main types of contributions (router, catalog, scaffolder, search, events, permissions) for both systems.
> > 
> > 
> > Would you be alright with limiting this and future discussion to just the new backend system? We'd really want to avoid inventing an additional intermediate mini backend system.
> 
> I understand this, it would definitely be possible to say new backend system is a pre-requisite here. But do we have any timeline when that will become the default?

We're actively working towards that right now. We want to ship some of the changes I mentioned later on first, as well as migrating some more existing plugins. Can't say when it's gonna be the default, but I'm pretty confident we'll be able to encourage use of it in 2-3 releases.

> OTOH it's not clear to me why the type of legacy backend support proposed by this RFC would necessarily lead to inventing or promoting an additional intermediate mini backend system.

This RFC proposes such a new system :sweat_smile:, specifically the `LegacyBackendPluginInstaller` and surrounding features.

> > > 5. Making a plugin compatible should require as **minimal and non-invasive source code changes** as possible.
> > > 6. It should be possible to repackage an existing backend plugin as a dynamic plugin **without requiring any change** in the existing plugin source code.
> > 
> > 
> > I assume you mean compared to using it as a regular backend plugin right? As in point 5 has already been done.
> 
> The point 6 is rather for situations where you want to create a dynamic plugin from a thirdparty non-dymanic one. It should be possible to write a wrapper.

Ah alright. As mentioned a bit further up we're hoping that existing plugins that use the new system will support declarative integration without any changes.

> > We're continuing the work by adding support for the design that you proposed through the `dynamicPluginsServiceFactory()`. Most likely it'll be an async loader of `BackendFeature`s. Our initial implementation will be targeted towards integrating packages that are already installed, but it's very much going to allow for you to experiment with dynamically loading packages at runtime as well! tada
> 
> The following Draft PR (that will hopefully be available for review very soon), implements such a loading of `BackendFeature`s in the context of dynamic plugins: #18862

Yep happy to add the `BackendInitializer` bits :grin:
 
> > Assuming all goes as planned we'll effectively consider declarative integration for backend a solved problem. There are a few more bits that we need ship as well, most importantly we're planning to remove the possibility to pass options to backend features through code, e.g. `catalogPlugin({ processingInterval: 30 })`, that'd have to be config instead - we didn't find any meaningful way to work that into the system.
> 
> That makes sense, though maybe it would be interesting to keep both options: allow providing these options through both configuration and code. Would it make sense ? I'm also thinking at things like the custom [KubernetesClustersSupplier](https://backstage.io/docs/reference/plugin-kubernetes-backend.kubernetesclusterssupplier): How would we pass this through configuration ? Or would this type of options always be managed through extensions?

Allowing both is the main alternative that we considered, but in the end it's just really confusing. As a plugin builder, when do you read things from top-level config vs options? It also doesn't work that well with our existing configuration schema system.

Things like `KubernetesClustersSupplier` have to be configured through extension points. It's already effectively the case since you don't have access to services when passing options. Only extremely simple callback-based options would make sense in their current form, and even for those it would still be best to have access to services.

**Comment by davidfestal** [comment]
> Quick reply to keep the discussion going grin
> 
> > > > 1. A dynamic plugin should be loaded through a standard dynamic `import(<path>)` call, which would return a module containing all the exported definitions of the related backstage plugin.
> > > 
> > > 
> > > Something we'll want to figure out here is the best way to access exports of said module. That goes for frontend as well, so a problem we should just solve once.
> > 
> > 
> > Why do you think we should need yet another way to access exports? The entrypoint serves that purpose. I wouldn’t connect this to the frontend system at all, loading additional assets over network is not even remotely related to loading code into the NodeJS runtime.
> 
> What I had in mind is how to select specific exports in configuration, or standardize on some specific export. In our experimentation what we're leaning towards at this point is to consider all exports of a module, filter out the ones that are backend features, and then install those. It's straightforward and doesn't require any changes to existing packages that already support the new backend system. On top of that we can declare ways to exclude or only include select exports from a package if that's needed as an escape hatch.

I get it. I agree that this would fit quite naturally as a followup of the current RFC, at least when support for the legacy backend system is not needed anymore.
This should probably be the end goal, but I assume that starting with the current approach (already implemented as an experimental package in PR #18862) for dynamic backend plugins would not be a problem, since the proposal of the current RFC (named entrypoint) is in fact a subset of what you propose (all available entry points that fit a given type).
BTW according to what I saw, it seems quite simple to integrate the work in PR #18862  with the implementation you propose in #18906. Both seem very complementary.

> 
> > > > 2. To make this loading both possible and robust, the plugin should be repackaged to be **standalone**, and **self-contained**, providing its specific dependencies
> > > 
> > > 
> > > Just wanna highlight that native dependencies are gonna be an annoyance that we need to figure out how to handle. My initial thinking here is that we make them external dependencies that plugins expect to be installed directly in the backend. To support multiple versions of the same package we could allow aliasing with overrides.
> > 
> > 
> > That doesn't seem realistic to me in terms of user experience, if I understood correctly what you mean here. Allowing such overrides in very specific cases, when the isolation of dynamic plugin dependencies isn't sufficient, could be useful, sure. But requiring the end-user to install all dynamic-plugin-specific dependencies himself, and provide such overrides to resolve himself all the possible version conflicts, doesn't seem to fit well with the user experience expected from the dynamic plugins feature.
> 
> Native modules as in ones that use [Node.js' C++ APIs](https://nodejs.org/api/addons.html). There's quite few of these but they may require explicit installation as opposed to just being shipped in a JS code bundle. One existing example is the [`isolated-vm`](https://www.npmjs.com/package/isolated-vm), which in addition to its own installation also for example requires the following packages to be installed on Ubuntu: `python g++ build-essential`.
> 

OK, sorry. I just didn't realize you were speaking of **such native modules** and not only of dynamic-plugin-specific private modules.

> It's just a technical issue, and if we can figure out a way to handle them in a way that doesn't require additional steps that's ofc preferable.

:100: 

> 
> > > > 3. As a special case, it should never provide any core `@backstage` dependency. On the contrary, it should always **share `@backstage` dependencies** with the backend application (to avoid having dupes of `@backstage` modules)
> > > 
> > > 
> > > Why? sweat_smile All relevant libraries should be built so support duplication, and I expect the cost of duplication here to be minimal. I think this is something we can leave as a potential future optimization.
> > 
> > 
> > There are always dependencies which are required to be singleton. E.g. a logger can’t open 2 file descriptors for the same log file. More generally, if we think of how runtime plugins would work in an extensible framework architecture, it seems to me that it makes sense to distinguish **plugin-specific private dependencies**, from **shared dependencies provided by the framework**. In the context of this RFC, I think it would make sense to consider **backstage dependencies as provided by the framework** by default. In NodeJS terms, this would mean: make them peer dependencies, and ensure that their packages are loaded from the same location as when they are loaded by the main backstage application.
> 
> This is already a problem we have to solve for for backends that are integrated through code. Un-solving it for dynamic plugins seems unnecessary.

The first, experimental implementation of this RFC, proposed in PR #18862, already solves it, so that the dynamic backend plugin feature could already be tested, played with, and refined on the current state of the backstage codebase.
But to make it clear, the current implementation solves it without making it a requirement of the `backend plugin manager` component itself, and of the loading mechanism. So it will be straightforward to remove this sharing constraint when not necessary anymore.

> When there is a strict requirement for a singleton it can be installed as an explicit service factory. Forcing sharing of certain dependencies adds a lot of complexity, because it makes us much more limited in our ability to remove exports. For example, with multiple versions of `@backstage/backend-plugin-api` allowed, we can remove or change exports such as `createBackendPlugin` in newer versions, but still keep support for older exports in `@backstage/backend-app-api`.

I agree that it should be the final goal.

> 
> > > > 4. A dynamic plugin should be loaded from either the old backend system, or the new backend system, according to the context, and provide contribution to the main types of contributions (router, catalog, scaffolder, search, events, permissions) for both systems.
> > > 
> > > 
> > > Would you be alright with limiting this and future discussion to just the new backend system? We'd really want to avoid inventing an additional intermediate mini backend system.
> > 
> > 
> > I understand this, it would definitely be possible to say new backend system is a pre-requisite here. But do we have any timeline when that will become the default?
> 
> We're actively working towards that right now. We want to ship some of the changes I mentioned later on first, as well as migrating some more existing plugins. Can't say when it's gonna be the default, but I'm pretty confident we'll be able to encourage use of it in 2-3 releases.

What about keeping the support for both backend systems in the first experimental implementation of the RFC (PR #18862) , since it is already available and doesn't involve more than alternate type definitions ?
We would obviously remove them when the switch to the new backend system is over.

> 
> > OTOH it's not clear to me why the type of legacy backend support proposed by this RFC would necessarily lead to inventing or promoting an additional intermediate mini backend system.
> 
> This RFC proposes such a new system sweat_smile, specifically the `LegacyBackendPluginInstaller` and surrounding features.

What I meant is that the `LegacyBackendPluginInstaller` is only the definition of an alternate type.
For the legacy backend, all usages of this type, and surrounding features would need to be implemented in the user backend application (as of today). And since the required code is not included in the main `packages/backend` application (PR #18862 adds such an example backend application in the `contrib` directory for testing purposes), I doubt that using this API as an intermediate mini backend system would really be a serious option for anyone.
On the contrary, the dynamic backend plugin support on the new backend system is straightforward and requires no structural change in the `backend-next` application.
Couldn't we keep the `LegacyBackendPluginInstaller` alternate API for now, on condition of making it clear, in comments and docs, that it will disappear at some point and should not be misused (even mark them deprecated) ?

> 
> > > > 5. Making a plugin compatible should require as **minimal and non-invasive source code changes** as possible.
> > > > 6. It should be possible to repackage an existing backend plugin as a dynamic plugin **without requiring any change** in the existing plugin source code.
> > > 
> > > 
> > > I assume you mean compared to using it as a regular backend plugin right? As in point 5 has already been done.
> > 
> > 
> > The point 6 is rather for situations where you want to create a dynamic plugin from a thirdparty non-dymanic one. It should be possible to write a wrapper.
> 
> Ah alright. As mentioned a bit further up we're hoping that existing plugins that use the new system will support declarative integration without any changes.
> 
> > > We're continuing the work by adding support for the design that you proposed through the `dynamicPluginsServiceFactory()`. Most likely it'll be an async loader of `BackendFeature`s. Our initial implementation will be targeted towards integrating packages that are already installed, but it's very much going to allow for you to experiment with dynamically loading packages at runtime as well! tada
> > 
> > 
> > The following Draft PR (that will hopefully be available for review very soon), implements such a loading of `BackendFeature`s in the context of dynamic plugins: #18862
> 
> Yep happy to add the `BackendInitializer` bits 😁
> 

Thanks !

> > > Assuming all goes as planned we'll effectively consider declarative integration for backend a solved problem. There are a few more bits that we need ship as well, most importantly we're planning to remove the possibility to pass options to backend features through code, e.g. `catalogPlugin({ processingInterval: 30 })`, that'd have to be config instead - we didn't find any meaningful way to work that into the system.
> > 
> > 
> > That makes sense, though maybe it would be interesting to keep both options: allow providing these options through both configuration and code. Would it make sense ? I'm also thinking at things like the custom [KubernetesClustersSupplier](https://backstage.io/docs/reference/plugin-kubernetes-backend.kubernetesclusterssupplier): How would we pass this through configuration ? Or would this type of options always be managed through extensions?
> 
> Allowing both is the main alternative that we considered, but in the end it's just really confusing. As a plugin builder, when do you read things from top-level config vs options? It also doesn't work that well with our existing configuration schema system.
> 
> Things like `KubernetesClustersSupplier` have to be configured through extension points. It's already effectively the case since you don't have access to services when passing options. Only extremely simple callback-based options would make sense in their current form, and even for those it would still be best to have access to services.

I agree.
